cores/esp8266/uart.c

```diff
@@ -212,6 +220,44 @@ uart_read_char(uart_t* uart)
     return data;
 }
 
+// loopback-test BW jumps by 190%
+size_t
+uart_read(uart_t* uart, char* userbuffer, size_t usersize)
```

Comment:
The main loop executes with interrupts disabled, so during the looping the uart isr won't trigger. The first part of the loop execution (i.e.: after the !uart_rx_buffer_available() if statement) is to empty the sw buffer, and while that is going on, the hw fifo will continue to fill. What is the worst case scenario here? Is it possible for the hw fifo to overrun in the time it takes to empty the sw buffer? Or is it a guarantee that the sw buffer will always be emptied faster than the time it takes for the hw fifo to overrun?
I'm not saying change the code, you are doing stress testing, after all. But this is something worth thinking about, and if it's not possible to guarantee that the hw fifo won't overrun, i.e.: with code time calculation and byte timing math, I think there should at least be a comment about the possibility here for future reference.

Reply:
Indeed interrupts are disabled so there is a risk the hw buffer gets overflowed. @mribble's previous patch which reduced the hw-rx-fifo interrupt trigger (hw buffer full of 127chars down to 100) is very good because it ensures that hw buffer has at least 127-100 free bytes at the moment when interrupts are disabled.
Maybe what could be good is to enable+disable interrupts after memcpy() in case the hw buffer gets too stuffed during transfer.
I need to enable and try  UIOF (hw-rx-fifo overflow) and try to trigger (and report) it.
Until now, I avoided time calculation by doing stress tests @3Mbauds (2343Kibits/s at most, reaching 2302Kibits/s), but this is not robust enough.